ec/src/pairing.rs

```diff
@@ -0,0 +1,201 @@
+pub struct PairingOutput<P: Pairing>(P::TargetField);
+
+impl<P: Pairing> PairingOutput<P> {
+    /// Converts `self` into an element of the underlying field.
+    pub fn to_field_element(&self) -> P::TargetField {
+        self.0
+    }
+}
+
+impl<P: Pairing> Display for PairingOutput<P> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl<P: Pairing> Zero for PairingOutput<P> {
+    /// The identity element, or "zero", of the group is the identity element of the multiplicative group of the underlying field, i.e., `P::TargetField::one()`.
+    fn zero() -> Self {
+        Self(P::TargetField::one())
+    }
+
+    fn is_zero(&self) -> bool {
+        self.0.is_one()
+    }
+}
+
+impl<'a, P: Pairing> Add<&'a Self> for PairingOutput<P> {
+    type Output = Self;
+
+    #[inline]
+    fn add(mut self, other: &'a Self) -> Self {
+        self += other;
+        self
+    }
+}
+
+impl<'a, P: Pairing> AddAssign<&'a Self> for PairingOutput<P> {
+    fn add_assign(&mut self, other: &'a Self) {
+        self.0 *= other.0;
+    }
+}
+
+impl<'a, P: Pairing> SubAssign<&'a Self> for PairingOutput<P> {
+    fn sub_assign(&mut self, other: &'a Self) {
+        self.0 *= other.0.inverse().unwrap();
+    }
+}
+
+impl<'a, P: Pairing> Sub<&'a Self> for PairingOutput<P> {
+    type Output = Self;
+
+    #[inline]
+    fn sub(mut self, other: &'a Self) -> Self {
+        self -= other;
+        self
+    }
+}
+
+ark_ff::impl_additive_ops_from_ref!(PairingOutput, Pairing);
+
+impl<P: Pairing> MulAssign<P::ScalarField> for PairingOutput<P> {
+    fn mul_assign(&mut self, other: P::ScalarField) {
+        *self = *self * other.into_repr()
+    }
+}
+
+impl<P: Pairing> Mul<<P::ScalarField as PrimeField>::BigInt> for PairingOutput<P> {
+    type Output = Self;
+    fn mul(mut self, other: <P::ScalarField as PrimeField>::BigInt) -> Self {
+        self.0 = self.0.pow(other.as_ref());
+        self
+    }
+}
+
+impl<P: Pairing> Zeroize for PairingOutput<P> {
+    fn zeroize(&mut self) {
+        self.0.zeroize()
+    }
+}
+
+impl<P: Pairing> Neg for PairingOutput<P> {
+    type Output = Self;
+
+    #[inline]
+    fn neg(self) -> Self {
+        Self(self.0.inverse().unwrap())
+    }
+}
+
+impl<P: Pairing> Distribution<PairingOutput<P>> for Standard {
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> PairingOutput<P> {
+        // Sample a random G1 element
+        let g1 = P::G1::rand(rng);
+        // Sample a random G2 element
+        let g2 = P::G2::rand(rng);
+        P::pairing(g1, g2)
+    }
+}
+
+impl<P: Pairing> Group for PairingOutput<P> {
+    type ScalarField = P::ScalarField;
+    type UniqueRepr = Self;
+
+    fn generator() -> Self::UniqueRepr {
+        // TODO: hardcode these values.
+        // Sample a random G1 element
+        let g1 = P::G1::generator();
+        // Sample a random G2 element
+        let g2 = P::G2::generator();
+        P::pairing(g1.into(), g2.into())
```

Comment:
To obtain the GT generator, we just pair the G1 and G2 generators. Since this value is fixed, we can technically hardcode it, but that requires work, and is also unlikely to really be used anywhere.

Reply:
https://indiv0.github.io/lazycell/lazycell/
https://doc.rust-lang.org/std/lazy/index.html

Reply:
ooh I wonder if we should use that for other APIs too now...

Reply:
Atomic laziness incurs a performance hit of like 2ns I hear, but afaik most protocols access their generator only rarely, so whatever.  It's more likely recomputation once per thread or once per exec causes slowness, especiall in WASM, so then the fastest would be pre-computed or if const generics work then 
```
const GENERATOR: Self::UniqueRepr = {
    let g1 = P::G1::generator();
    let g2 = P::G2::generator();
    P::pairing(g1.into(), g2.into())
};
```
but you could just leave a performance TODO for whoever cares about those cases.

As an aside, we've layers of tables that depend upon tables in https://github.com/w3f/rs-ec-perf/ so I used https://gitlab.com/okannen/static_init to avoid the 2ns performance hit, but it brings portability issues so likely unsuitable here.  As a further aside, rand had a long discussion around mutable thread locals in https://github.com/rust-random/rand/issues/968